### PR TITLE
enhance: ask user for OpenAI key and store it in the cred store

### PR DESCRIPTION
**Pull request by g-linville**

### CODE REVIEW

**Review by thedadams — APPROVED** [review]

File: pkg/gptscript/gptscript.go [hunk 1]
```diff
@@ -65,7 +66,12 @@ func New(opts *Options) (*GPTScript, error) {
 		return nil, err
 	}
 
-	oAIClient, err := openai.NewClient(opts.OpenAI, openai.Options{
+	cliCfg, err := config.ReadCLIConfig(opts.OpenAI.ConfigFile)
+	if err != nil {
+		return nil, err
+	}
+
+	oAIClient, err := openai.NewClient(cliCfg, opts.CredentialContext, opts.OpenAI, openai.Options{
```
Comment (njhale) [inline pkg/gptscript/gptscript.go:74]: nit:

```suggestion
	oaiClient, err := openai.NewClient(cliCfg, opts.CredentialContext, opts.OpenAI, openai.Options{
```

**Review by njhale — APPROVED** [review]

**Comment by g-linville** [comment]
I made some pretty substantial changes to this PR:

- We now use SysPrompt to ask the user (which means I had to move it to its own package in order to avoid an import cycle)
- We only prompt the user for the token when it is actually needed
  - This involved creating a custom error type and checking for it in a couple places, which is a bit gross but is the cleanest solution I could come up with for this

File: pkg/prompt/prompt.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	data, err = io.ReadAll(resp.Body)
+	return string(data), err
+}
+
+func SysPrompt(ctx context.Context, envs []string, input string) (_ string, err error) {
+	var params struct {
+		Message   string `json:"message,omitempty"`
+		Fields    string `json:"fields,omitempty"`
+		Sensitive string `json:"sensitive,omitempty"`
+	}
```
Comment (StrongMonkey) [inline pkg/prompt/prompt.go:48]:
```suggestion
	var params types.Prompt
```

Comment (StrongMonkey) [reply pkg/prompt/prompt.go:48]: Looks like this can reuse the existing types.Prompt

**Review by StrongMonkey — APPROVED** [review]

File: pkg/prompt/prompt.go (hunk 2 above)
Comment (StrongMonkey) [reply pkg/prompt/prompt.go:48]: Oh I see, then nvm

**Comment by cjellick** [comment]
We will need to make sure this works with @tylerslaton's UI, once he has prompting support

File: pkg/credentials/credential.go [hunk 3]
```diff
@@ -8,9 +8,17 @@ import (
 	"github.com/docker/cli/cli/config/types"
 )
 
+type CredentialType string
+
+const (
+	CredentialTypeTool          CredentialType = "tool"
+	CredentialTypeModelProvider CredentialType = "modelprovider"
```
Comment (StrongMonkey) [inline pkg/credentials/credential.go:15]: Nit: `modelProvider` might be better?

File: pkg/credentials/helper.go [hunk 4]
```diff
@@ -55,7 +57,32 @@ func (h *HelperStore) GetAll() (map[string]types.AuthConfig, error) {
 		return nil, err
 	}
 
-	for serverAddress := range serverAddresses {
+	newCredAddresses := make(map[string]string, len(serverAddresses))
+	for serverAddress, val := range serverAddresses {
```
Comment (StrongMonkey) [inline pkg/credentials/helper.go:61]: Interesting... that's weird behavior from mac library

File: pkg/credentials/credential.go [hunk 5]
```diff
@@ -33,14 +41,28 @@ func credentialFromDockerAuthConfig(authCfg types.AuthConfig) (Credential, error
...
+	// We used to hardcode the username as "gptscript" before CredentialType was introduced, so
+	// check for that here.
+	credType := authCfg.Username
+	if credType == "gptscript" {
+		credType = string(CredentialTypeTool)
+	}
+
+	// If it's a tool credential or sys.openai, remove the http[s] prefix.
+	address := authCfg.ServerAddress
+	if credType == string(CredentialTypeTool) || strings.HasPrefix(address, "https://sys.openai///") {
```
Comment (StrongMonkey) [inline pkg/credentials/credential.go:53]: Is `///` a typo or is it desired to be this way?

**Review by StrongMonkey — APPROVED** [review]

File: pkg/credentials/helper.go (hunk 4 above)
Comment (g-linville) [reply pkg/credentials/helper.go:61]: Yep it definitely took me a bit to figure that one out lol

File: pkg/credentials/credential.go (hunk 3 above)
Comment (g-linville) [reply pkg/credentials/credential.go]: fixed

File: pkg/credentials/credential.go (hunk 5 above)
Comment (g-linville) [reply pkg/credentials/credential.go]: It was intentional. I use it as the separator between the credential tool name/alias and the credential context, both of which get stored in the "server address" of the credential in the keychain. I created a const for this to add more clarity.

File: pkg/openai/client.go [hunk 6]
```diff
@@ -121,8 +162,9 @@ func (c *Client) ListModels(ctx context.Context, providers ...string) (result []
 		return nil, nil
 	}
 
+	// If auth is invalid, we just want to return nothing.
```
Comment (thedadams) [inline pkg/openai/client.go:165]: nit: could you amend the comment to explain why we want to do nothing?

**Review by thedadams — APPROVED** [review]

File: pkg/openai/client.go (hunk 6 above)
Comment (g-linville) [reply pkg/openai/client.go:165]: Done

**Review by thedadams — APPROVED** [review]

**Pull request merged**